upgrade.go

```diff
@@ -118,7 +135,30 @@ func (b *versionedKVBackend) Upgrade(ctx context.Context, s logical.Storage) err
 		return nil
 	}
 
-	upgradeInfo := &UpgradeInfo{
+	// See if we're already done, in which case we just need to ensure the
+	// policy was created
+	upgradeEntry, err := s.Get(ctx, path.Join(b.storagePrefix, "upgrading"))
+	if err != nil {
+		return err
+	}
+
+	upgradeInfo := new(UpgradeInfo)
+	if upgradeEntry != nil {
+		err := proto.Unmarshal(upgradeEntry.Value, upgradeInfo)
+		if err != nil {
+			return err
+		}
+	}
+
+	if upgradeInfo.Done {
+		// Just synchronously call policy
+		if _, err = b.policy(ctx, s); err != nil {
```

Comment:
I see a warning about "The caller must have the backend lock." in `policy()`. Applies in the upgrade case too?

Reply:
That's a good point -- It's fine in the non-goroutine case because it's synchronous during backend factory creation. But if we don't run it then, and end up running it in the goroutine, then we don't have the lock at that time. This the presents an issue, because if we aren't upgraded, if we write the policy it'll be considered by the upgrade logic to be existing data and will be moved. So it has to be done after upgrading. But in the meantime clients can make calls that can then indirectly call policy. So I may indeed have to wrap the second invocation (in the goroutine) in a lock and it's just best effort. That said since secondaries will wait for the write of the upgrade done canary anyways probably they'll block client access until then so it'll be fine just moving that block up a bit.